### PR TITLE
Trie proof API refactor

### PR DESCRIPTION
**Pull request by ScottyPoi**
This PR improves upon the proof methods in the `Trie` library.

The Trie methods  `fromProof` and `verifyProof` require a new Trie object to be instantiated by the user simply for the purpose of calling the methods.  This can be confusing, and seems unnecessarily indirect.

With static versions of the same methods, users can now accomplish these tasks in one step.

1. `Trie.fromProof`
- Works just like `trie.fromProof`, with an additional `opts` parameter to pass in keyHashing settings.

Now instead of:
```
const trie = new Trie(opts)
await trie.fromProof(proof)
```
We can just do:
```
const trie = await Trie.fromProof(proof, opts)
```

2. `Trie.verifyProof`
- Works just like `trie.verifyProof`, with an additional `opts` parameter to pass in keyHashing settings.
- Does not require the `rootHash` parameter as expected by class method.

Now instead of:
```
const trie = new Trie(opts)
const val = trie.verifyProof(trie.root(), key, proof, opts)
```
We can just do:
```
const val = Trie.verifyProof(key, proof, opts)
```

### CODE REVIEW

**Comment by scorbajio** [comment]
Linking #2927

File: packages/trie/src/trie.ts
Comment (jochem-brouwer) [inline packages/trie/src/trie.ts:116]: What if I want to import two proofs? Or, I want to import a new proof in my trie later? This is now not possible anymore, right?

**Review by jochem-brouwer — COMMENTED** [review]
I'm not sure if we want to make `fromProof` static

File: packages/trie/src/trie.ts
Comment (holgerd77) [reply packages/trie/src/trie.ts:116]: Ok, let's bring this back to the drawing board and first decide if this is a change we want respectively which makes sense and is thought through to the end.

Comment (holgerd77) [reply packages/trie/src/trie.ts:116]: Will give this a "blocked" label for now.

Comment (jochem-brouwer) [reply packages/trie/src/trie.ts:116]: (this change (i.e. make `fromProof` static) does not seem to be referred in the linked issue)

Comment (scorbajio) [reply packages/trie/src/trie.ts:116]: > What if I want to import two proofs? Or, I want to import a new proof in my trie later? This is now not possible anymore, right?

It seems like these tasks are made possible with the new `updateFromProof` function.

Comment (jochem-brouwer) [reply packages/trie/src/trie.ts:116]: Ah yes you are right, I somehow missed that.

Comment (holgerd77) [reply packages/trie/src/trie.ts:116]: Feel free to "unblock" this again or course of you guys think you've gotten answers for most of the questions here and things can be introduced in a non breaking way, you guys have a lot better overview than me here what's useful and makes sense regarding all this proof stuff!

File: packages/trie/src/trie.ts
Comment (jochem-brouwer) [inline packages/trie/src/trie.ts:819]: I am not sure if I like this check, have to think about the practical downsides of this :thinking:

**Review by scorbajio — CHANGES_REQUESTED** [review]

File: packages/trie/src/trie.ts
Comment (ScottyPoi) [reply packages/trie/src/trie.ts:116]: yes, `updateFromProof` is just the old `fromProof` with the word `update` tagged on to distinguish from the other.  To avoid actual breaking changes, we could instead use `createFromProof` for the static method, and leave `fromProof` as it was.

File: packages/trie/src/trie.ts
Comment (ScottyPoi) [reply packages/trie/src/trie.ts:819]: If you try to update a sparse trie with an invalid proof, this will throw.
shift() because the root node is already in the trie, so we'll never need to add it.

File: packages/trie/src/trie.ts
Comment (ScottyPoi) [reply packages/trie/src/trie.ts:146]: Three, actually, because the first node in the proof is also the root node.  I didn't write the method, I just made it static, and left it mostly the way it was.  The top `rootHash` parameter is definitely not needed, and the "opts" here really just needs to set the `keyHashing` behavior.

File: packages/statemanager/src/stateManager.ts
Comment (ScottyPoi) [reply packages/statemanager/src/stateManager.ts:724]: agreed

**Comment by ScottyPoi** [comment]
Should be non-breaking now :egg:

File: packages/trie/src/trie.ts
Comment (ScottyPoi) [reply packages/trie/src/trie.ts]: :thinking:  I'd have to hear the upside to allowing a trie to update itself from a proof to a different trie!

File: packages/client/test/sim/beaconsync.md
Comment (jochem-brouwer) [inline packages/client/test/sim/beaconsync.md:1]: Random changes in this file?

File: packages/client/test/testdata/blocks/kaustinen2.json
Comment (jochem-brouwer) [inline packages/client/test/testdata/blocks/kaustinen2.json:1]: Also random changes here?

File: packages/statemanager/src/rpcStateManager.ts
Comment (jochem-brouwer) [inline packages/statemanager/src/rpcStateManager.ts:13]: These comments should be deleted

File: packages/trie/src/trie.ts
Comment (jochem-brouwer) [reply packages/trie/src/trie.ts:175]: We now have https://github.com/ethereumjs/ethereumjs-monorepo/pull/3186. I think we want to make the methods static in some cases? (But this is breaking)

**Review by jochem-brouwer — COMMENTED** [review]
I am not entirely sure about the state of this PR. We now have https://github.com/ethereumjs/ethereumjs-monorepo/pull/3186, it seems that if we want this in this seems to be merely marking some methods as `static`? (Note: we should make this non-breaking, so it should both be static and non-static). I will remove the `blocked` label until we resolve what we want with this PR. (I think the gist is to make the methods static which should have been static in the first place? :thinking: )

**Comment by holgerd77** [comment]
> I will remove the blocked label until we resolve what we want with this PR

@jochem-brouwer I assume that you intended to write "I will not remove..." if not otherwise stated since that would make more sense 🙂 

@scorbajio don't know, are you getting along with the things we get from #3186 ? If so, @jochem-brouwer: would you then recommend to close this PR and rather not merge?

**Comment by jochem-brouwer** [comment]
Yes indeed I did intend to write: I will not remove :)

I think that this PR can be updated with the changes of #3186, but I think we should update this PR to also include small changes such as also making `verifyProof` static (besides keeping it non-static, because otherwise it is breaking). So that would change this PR to a few lines of changes.

( @holgerd77 I think you intended to ping @ScottyPoi in your last post)

**Comment by holgerd77** [comment]
@jochem-brouwer 👍

Ah, no, I indeed intended to mention @scorbajio since he requested this work to be updated and reviewed for some other follow up work.

File: packages/client/test/sim/beaconsync.md
Comment (scorbajio) [reply packages/client/test/sim/beaconsync.md:1]: Must have been from one of the merges. I fixed it so it's just the single extra space removed and any other unnecessary changes excluded.

File: packages/client/test/testdata/blocks/kaustinen2.json
Comment (scorbajio) [reply packages/client/test/testdata/blocks/kaustinen2.json:1]: Fixed.

File: packages/trie/src/trie.ts
Comment (scorbajio) [reply packages/trie/src/trie.ts:180]: Yes, the goal is to have a static version of `verifyProof` so that a new trie instantiation isn't needed for each proof verification, or even more awkward is the proofTrie pattern where we keep a trie around in both the statemanager and snap sync fetchers to perform proof verification.

The changes fro #3186 should be included, so we should have a static and non-static version of `verifyProof` and `fromProof`. This shouldn't be breaking, right? Since users can just use the previously available non-static versions if they chose to.

**Comment by scorbajio** [comment]
@jochem-brouwer There is currently both a static and non-static version of both `verifyProof` and `fromProof` available, but it might not show up since it's been rebased on `master`, so it should no longer be breaking, right? I've updated the function docstrings to make this more visible.

**Comment by holgerd77** [comment]
I've completely lost track here who wants what.  😋 

It would nevertheless be good if we bring this conversation and PR to a close.

Some things to keep in mind: we should both...

a) work towards a clean API and
b) remain backwards compatible and
c) see how we can fullfil the needs from Amir in some way (either he needs some changes or his code can be adjusted)

File: packages/trie/src/trie.ts
Comment (jochem-brouwer) [inline packages/trie/src/trie.ts:195]: This code is now redundant, can this call into `trie.fromProof`? The `if (opts?.root && !equalsBytes(opts.root, trie.hash(proof[0]))) {` check should stay though.

However, another question then here is, we have `createTrieFromProof` which almost does exactly the same, except that this `fromProof` has the extra check that the `trie.hash(proof[0])` matches the `opts.root` :thinking: (if it is provided :thinking: )

**Review by scorbajio — COMMENTED** [review]
LGTM!

File: packages/trie/src/trie.ts
Comment (jochem-brouwer) [reply packages/trie/src/trie.ts:245]: I think this one needs to be checked @scorbajio

File: packages/trie/src/trie.ts
Comment (scorbajio) [reply packages/trie/src/trie.ts:150]: Added.

File: packages/client/src/sync/fetcher/accountfetcher.ts
Comment (scorbajio) [reply packages/client/src/sync/fetcher/accountfetcher.ts:331]: Good point.

File: packages/trie/src/trie.ts
Comment (scorbajio) [reply packages/trie/src/trie.ts:244]: I've switched it over to `updateFromProof` to remove redundancies.

**Comment by holgerd77** [comment]
Updated this via UI

I think we can take this in after having done a rough review? Will wait for a confirmation from @jochem-brouwer though (then optimally directly merge).

**Review by jochem-brouwer — APPROVED** [review]
Resolved merge conflicts and fixed the example. LGTM (assuming CI passes, will directly merge)

**Pull request merged**
